moveit_ros/planning/rdf_loader/src/synchronized_string_parameter.cpp

```diff
@@ -0,0 +1,154 @@
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: David V. Lu!! */
+
+#include <moveit/rdf_loader/synchronized_string_parameter.h>
+
+namespace rdf_loader
+{
+std::string SynchronizedStringParameter::loadInitialValue(const std::shared_ptr<rclcpp::Node>& node,
+                                                          const std::string& name, StringCallback parent_callback)
+{
+  node_ = node;
+  name_ = name;
+  parent_callback_ = parent_callback;
+
+  if (getMainParameter())
+  {
+    if (shouldPublish())
+    {
+      // Transient local is similar to latching in ROS 1.
+      string_publisher_ = node_->create_publisher<std_msgs::msg::String>(name_, rclcpp::QoS(1).transient_local());
+
+      std_msgs::msg::String msg;
+      msg.data = content_;
+      string_publisher_->publish(msg);
+    }
+    return content_;
+  }
+
+  // Load topic parameters
+  std::string keep_open_param = name_ + "_continuous";
+  if (!node_->has_parameter(keep_open_param))
+  {
+    node_->declare_parameter(keep_open_param, rclcpp::ParameterType::PARAMETER_BOOL);
+  }
+  bool keep_open;
+  node_->get_parameter_or(keep_open_param, keep_open, false);
+
+  std::string timeout_param = name_ + "_timeout";
+  if (!node_->has_parameter(timeout_param))
+  {
+    node_->declare_parameter(timeout_param, rclcpp::ParameterType::PARAMETER_DOUBLE);
+  }
+  double d_timeout;
+  node_->get_parameter_or(timeout_param, d_timeout, 10.0);  // ten second default
+  rclcpp::Duration timeout = rclcpp::Duration::from_seconds(d_timeout);
+
+  if (!waitForMessage(timeout))
+  {
+    RCLCPP_ERROR_ONCE(node_->get_logger(),
+                      "Could not find parameter %s and did not recieve %s via std_msgs::msg::String subscription "
+                      "within %f seconds.",
+                      name_.c_str(), name_.c_str(), d_timeout);
+  }
+  if (!keep_open)
+  {
+    string_subscriber_.reset();
+  }
+  return content_;
+}
+
+bool SynchronizedStringParameter::getMainParameter()
+{
+  // Check if the parameter is declared, declare it if it's not declared yet
+  if (!node_->has_parameter(name_))
+  {
+    node_->declare_parameter(name_, rclcpp::ParameterType::PARAMETER_STRING);
+  }
+
+  node_->get_parameter_or(name_, content_, std::string());
+
+  return !content_.empty();
+}
+
+bool SynchronizedStringParameter::shouldPublish()
+{
+  std::string publish_param = "publish_" + name_;
+  bool publish_string;
+  if (!node_->has_parameter(publish_param))
+  {
+    node_->declare_parameter(publish_param, rclcpp::ParameterType::PARAMETER_BOOL);
+  }
+  node_->get_parameter_or(publish_param, publish_string, false);
+  return publish_string;
+}
+
+bool SynchronizedStringParameter::waitForMessage(const rclcpp::Duration timeout)
+{
+  string_subscriber_ = node_->create_subscription<std_msgs::msg::String>(
+      name_, rclcpp::QoS(1).transient_local().reliable(),
```

Comment:
Same question here with the QoS.

Reply:
If change the subscriber to use `rclcpp::ParametersQoS()`, the QoS does not match with the robot state publisher (linked in https://github.com/ros-planning/moveit2/pull/765#discussion_r759344079). 

RSP is publishing with `Durability: TRANSIENT_LOCAL` and `rclcpp::ParametersQoS()` uses `Durability: VOLATILE`

```
! moveit2/ > ros2 topic info /robot_description -v
Type: std_msgs/msg/String

Publisher count: 1

Node name: robot_state_publisher
Node namespace: /
Topic type: std_msgs/msg/String
Endpoint type: PUBLISHER
GID: 49.c3.10.01.fb.a5.9d.f1.1c.23.b6.ee.00.00.17.03.00.00.00.00.00.00.00.00
QoS profile:
  Reliability: RELIABLE
  Durability: TRANSIENT_LOCAL
  Lifespan: Infinite
  Deadline: Infinite
  Liveliness: AUTOMATIC
  Liveliness lease duration: Infinite

Subscription count: 2

Node name: move_group
Node namespace: /
Topic type: std_msgs/msg/String
Endpoint type: SUBSCRIPTION
GID: 29.5c.10.01.bf.0e.05.00.a0.49.dd.7e.00.00.15.04.00.00.00.00.00.00.00.00
QoS profile:
  Reliability: RELIABLE
  Durability: VOLATILE
  Lifespan: Infinite
  Deadline: Infinite
  Liveliness: AUTOMATIC
  Liveliness lease duration: Infinite
```


So I see some options:

  0. Add a QoS parameter to `rdf_loader` and `SynchronizedStringParameter`
  0. Leave it as-is, making `robot_description` and `robot_description_semantic` always have the same QoS (matching RSP)
  0. Try to change RSP to use `ParametersQoS`

I favor option 1 (leaving as-is) because its simpler and I don't care to learn the differences between the QoS settings. What do you think @tylerjw ?

Reply:
I didn't realize that this QoS was already specified in robot_state_publisher.  In that case, I am in favor of leaving it as is (option 1) and matching the QoS from robot_state_publisher.